python/ql/test/library-tests/frameworks/lxml/taint_test.py

```diff
@@ -0,0 +1,154 @@
+import lxml.etree as ET
+import io
+import typing
+
+def ensure_tainted(*args):
+    print("ensure_tainted: ", *args)
+
+TAINTED_STRING = "<a><b></b><c></c></a>"
+src = TAINTED_STRING
+
+def test():
+    ensure_tainted(
+        src, # $ tainted
+        ET.fromstring(src),  # $ tainted decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.fromstring(..)
+        ET.XML(src),  # $ tainted decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.XML(..)
+        ET.HTML(src),  # $ tainted decodeFormat=XML decodeInput=src decodeOutput=ET.HTML(..)
+        ET.fromstringlist([src]),  # $ tainted decodeFormat=XML decodeInput=List xmlVuln='XXE' decodeOutput=ET.fromstringlist(..)
+        ET.XMLID(src),  # $ tainted decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.XMLID(..)
+        ET.XMLDTDID(src),  # $ tainted decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.XMLDTDID(..)
+    )
+
+
+    parser = ET.XMLParser()
+    parser.feed(src) # $ decodeFormat=XML decodeInput=src xmlVuln='XXE'
+    ensure_tainted(parser.close()),  # $ tainted decodeOutput=parser.close()
+
+    parser2 = ET.get_default_parser()
+    parser2.feed(data=src) # $ decodeFormat=XML decodeInput=src xmlVuln='XXE'
+    ensure_tainted(parser2.close()),  # $ tainted decodeOutput=parser2.close()
+
+    elem = ET.XML(src) # $ decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.XML(..)
+    ensure_tainted(
+        elem,  # $ tainted
+        ET.tostring(elem),  # $ tainted encodeFormat=XML encodeInput=elem encodeOutput=ET.tostring(..)
+        ET.tostringlist(elem),  # $ tainted encodeFormat=XML encodeInput=elem encodeOutput=ET.tostringlist(..)
+        ET.tounicode(elem), # $ tainted encodeFormat=XML encodeInput=elem encodeOutput=ET.tounicode(..)
+        elem.attrib,  # $ tainted
+        elem.base,  # $ tainted
+        elem.nsmap,  # $ tainted
+        elem.prefix,  # $ tainted
+        elem.tag,  # $ tainted
+        elem.tail,  # $ tainted
+        elem.text,  # $ tainted
+        elem[0],  # $ tainted
+        elem[0].text,  # $ tainted
+        elem.cssselect("b"),  # $ tainted
+        elem.cssselect("b")[0].text,  # $ tainted
+        elem.find("b").text,  # $ tainted
+        elem.findall("b"),  # $ tainted
+        elem.findall("b")[0].text,   # $ tainted 
+        list(elem.findall("b"))[0].text,  # $ MISSING:tainted # Type tracking is not followed through call to `list`, 
+        elem.get("at"),  # $ tainted
+        elem.getchildren()[0].text,  # $ tainted
+        elem.getiterator(),  # $ tainted
+        next(elem.getiterator()).text,  # $ MISSING:tainted
```

Comment:
`next` is currently modeled as a flow summary, but not a "simple" one, so we cannot type track through it (it uses type tracking to identify the calls).

Reply:
Is there anything we could expect to be able to type track through these cases of methods returning iterators?
(`list(X)[0]` doesn't work; and neither do for loops as one of the other tests here shows)

Reply:
Hm, it looks like all uses of `getACallSimple` are in test summaries, so probably not..